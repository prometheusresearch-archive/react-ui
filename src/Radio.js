/**
 * @copyright 2016-present, Prometheus Research, LLC
 */


import RadioBase from './RadioBase';
import Block from './Block';
import {style, css} from './stylesheet';
import {margin, textColors} from './theme';

export default style(RadioBase, {
  Label: {
    fontSize: '0.875rem',
    fontWeight: 400,
    color: '#444',
    disabled: {
      cursor: 'not-allowed',
      color: textColors.disabled,
    },
  },
  Input: {
    verticalAlign: 'middle',
    disabled: {
      cursor: 'not-allowed',
    },
  },
  LabelWrapper: style(Block, {
    cursor: css.cursor.default,
    display: 'inline-block',
    verticalAlign: 'middle',
    userSelect: 'none',
    ltr: {
      marginLeft: margin['x-small'],
    },
    rtl: {
      marginRight: margin['x-small'],
    },
  }),
  Hint: {
    fontSize: '0.6rem',
    disabled: {
      cursor: 'not-allowed',
      color: textColors.disabled,
    },
  }
});
