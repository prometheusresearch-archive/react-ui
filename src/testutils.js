/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';
import {isEqual} from 'lodash/lang';
import indent from 'indent-string';

import {
  findWithType as _findWithType,
  findAllWithType as _findAllWithType,
  getMountedInstance,
  findAll
} from 'react-shallow-testutils';

import {
  createRenderer as _createRenderer
} from 'react-addons-test-utils';

import _inspectElement, {
  inspectReactProp as _inspectReactProp
} from 'inspect-react-element';


import assert from 'power-assert';
import {stub, spy, mock} from 'sinon';

export {stub, spy, mock, assert};

class FormDataMock {

  append = spy();
}

let originalFormData = global.FormData;

export function mockFormData() {
  global.FormData = FormDataMock;
}

export function unmockFormData() {
  global.FormData = originalFormData;
}

function inspectReactProps(node) {
  return '__originalConfig' in node.props ?
    {...node.props.__originalConfig, children: node.props.children} :
    node.props || {};
}

function inspectReactProp(propName, propValue) {
  if (propName === '__originalConfig' || propName === 'key' || propName === 'ref') {
    return null;
  }
  return _inspectReactProp(propName, propValue);
}

export function inspectElement(node) {
  return _inspectElement(node, {inspectReactProps, inspectReactProp});
}

class ReactShallowRenderer {

  constructor() {
    this._renderer = _createRenderer();
  }

  render(element, context) {
    return this._renderer.render(element, context);
  }

  get instance() {
    return getMountedInstance(this._renderer);
  }

  get element() {
    return this._renderer.getRenderOutput();
  }

  findWithType(type) {
    return findWithType(this.element, type);
  }

  findAllWithType(type) {
    return findAllWithType(this.element, type);
  }

  assertElementWithType(type) {
    this.findWithType(type);
  }

  assertNoElementWithType(type) {
    assert(this.findAllWithType(type).length === 0);
  }


  findWithTypeProps(type, props) {
    return findWithTypeProps(this.element, type, props);
  }

  findAllWithTypeProps(type, props) {
    return findAllWithTypeProps(this.element, type, props);
  }

  assertElementWithTypeProps(type, props) {
    this.findWithTypeProps(type, props);
  }

  assertNoElementWithTypeProps(type, props) {
    assert(this.findAllWithTypeProps(type, props).length === 0);
  }


  findWithElement(element) {
    return findWithElement(this.element, element);
  }

  findAllWithElement(element) {
    return findAllWithElement(this.element, element);
  }

  assertElement(element) {
    this.findWithElement(element);
  }

  assertNoElement(element) {
    assert(this.findAllWithElement(element).length === 0);
  }

}

export function createRenderer() {
  return new ReactShallowRenderer();
}

export function findWithType(root, type) {
  let elements = findAll(root, _makeMatchWithType(type));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one partial match for element:
${indent(inspectElement(React.createElement(type)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithType(root, type) {
  return findAll(root, _makeMatchWithType(type));
}

function _makeMatchWithType(type) {
  return function _matchWithType(element) {
    return element && element.type && (
      element.type === type ||
      element.type.Component === type
    );
  };
}

export function findWithTypeProps(root, type, props) {
  let elements = findAll(root, _makeMatchWithElement(type, props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(React.createElement(type, props)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithTypeProps(root, type, props) {
  return findAll(root, _makeMatchWithElement(type, props));
}

export function findWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  let elements = findAll(root, _makeMatchWithElement(spec.type, spec.props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(spec), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  return findAll(root, _makeMatchWithElement(spec.type, spec.props));
}

function _makeMatchWithElement(type, props) {
  let matchWithType = _makeMatchWithType(type);

  let matchWithProps = function(element) {
    for (let key in props) {
      if (!props.hasOwnProperty(key)) {
        continue;
      }
      if (!element.props.hasOwnProperty(key)) {
        continue;
      }
      if (key === '__originalConfig') {
        continue;
      }
      if (!isEqual(element.props[key], props[key])) {
        return false;
      }
    }
    return true;
  };
  return function _matchWithElement(element) {
    return element && matchWithType(element) && matchWithProps(element);
  };
}

function _normalizeReactElementSpec(spec) {
  let {props: {__originalConfig = null}} = spec;
  if (__originalConfig) {
    return {...spec, props: __originalConfig};
  } else {
    return spec;
  }
}

let _createElement = React.createElement;

function createElement(type, config, ...children) {
  config = {...config, __originalConfig: config || {}};
  let element = _createElement(type, config, ...children);
  return element;
}

React.createElement = createElement;

