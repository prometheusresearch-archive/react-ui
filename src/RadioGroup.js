/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';

import Radio from './Radio';

export default class RadioGroup extends React.Component {

  static stylesheet = {
    Root: 'div',
    RadioWrapper: 'div',
    Radio: Radio,
  };

  render() {
    let {options} = this.props;
    let {Root} = this.constructor.stylesheet;
    options = options.map(this.renderOption, this);
    return (
      <Root>
        {options}
      </Root>
    );
  }

  renderOption(option) {
    let {RadioWrapper, Radio} = this.constructor.stylesheet;
    let value = this.props.value;
    let checked = value === option.id;
    return (
      <RadioWrapper key={option.id}>
        <Radio
          value={checked}
          label={option.title}
          onChange={this.onChange.bind(this, option.id)}
          />
      </RadioWrapper>
    );
  }

  onChange(id, checked) {
    if (checked) {
      this.props.onChange(id);
    }
  }
}

