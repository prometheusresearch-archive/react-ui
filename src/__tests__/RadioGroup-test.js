/**
 * @copyright 2016, Prometheus Research, LLC
 */

import React from 'react';

import {createRenderer, spy, assert} from '../testutils';
import Radio from '../Radio';
import RadioGroup from '../RadioGroup';

describe('react-ui', function() {

  let options = [
    {value: 'a', label: 'A'},
    {value: 'b', label: 'B'}
  ];

  let renderer = null;

  beforeEach(function() {
    renderer = createRenderer();
  });

  describe('<RadioGroup />', function() {

    it('renders', function() {
      let onChange = spy();
      renderer.render(
        <RadioGroup
          value={'a'}
          onChange={onChange}
          options={options}
          />
      );
      let radios = renderer.findAllWithType(Radio);

      assert(radios.length === 2);
      assert(radios[0].props.value);
      assert(!radios[1].props.value);

      radios[1].props.onChange(true);
      assert(onChange.callCount === 1);
      assert.deepEqual(onChange.lastCall.args[0], 'b');
    });
  });

});

