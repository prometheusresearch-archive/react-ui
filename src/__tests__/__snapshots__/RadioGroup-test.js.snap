// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="div-1648223554"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  role="radiogroup"
  tabIndex={0}
>
  <div
    className="div-828019517 div-vertical-828019517"
  >
    <div
      title={undefined}
    >
      <input
        aria-labelledby="aria1"
        checked={true}
        className="Input-161314557"
        disabled={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        tabIndex={-1}
        type="radio"
      />
      <div
        className="Block-2087937970 Block-ltr-2087937970"
        onClick={[Function]}
        style={
          Object {
            "bottom": undefined,
            "display": undefined,
            "float": undefined,
            "height": undefined,
            "left": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "maxHeight": undefined,
            "maxWidth": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "paddingBottom": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "position": "relative",
            "right": undefined,
            "textAlign": undefined,
            "top": undefined,
            "verticalAlign": undefined,
            "whiteSpace": undefined,
            "width": undefined,
          }
        }
      >
        <div
          className="div-3928652604"
          id="aria1"
        >
          A
        </div>
        <div
          className="div-2120500968"
        />
      </div>
    </div>
  </div>
  <div
    className="div-828019517 div-vertical-828019517"
  >
    <div
      title={undefined}
    >
      <input
        aria-labelledby="aria2"
        checked={false}
        className="Input-161314557"
        disabled={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        tabIndex={-1}
        type="radio"
      />
      <div
        className="Block-2087937970 Block-ltr-2087937970"
        onClick={[Function]}
        style={
          Object {
            "bottom": undefined,
            "display": undefined,
            "float": undefined,
            "height": undefined,
            "left": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "maxHeight": undefined,
            "maxWidth": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "paddingBottom": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "position": "relative",
            "right": undefined,
            "textAlign": undefined,
            "top": undefined,
            "verticalAlign": undefined,
            "whiteSpace": undefined,
            "width": undefined,
          }
        }
      >
        <div
          className="div-3928652604"
          id="aria2"
        >
          B
        </div>
        <div
          className="div-2120500968"
        />
      </div>
    </div>
  </div>
</div>
`;
